hbase-backup/src/test/java/org/apache/hadoop/hbase/backup/TestBackupRestoreWithModifications.java

```diff
@@ -0,0 +1,339 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.backup;
+
+import static org.apache.hadoop.hbase.backup.BackupInfo.BackupState.COMPLETE;
+import static org.apache.hadoop.hbase.backup.BackupType.FULL;
+import static org.apache.hadoop.hbase.backup.BackupType.INCREMENTAL;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.LocatedFileStatus;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.RemoteIterator;
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.HBaseClassTestRule;
+import org.apache.hadoop.hbase.HBaseConfiguration;
+import org.apache.hadoop.hbase.KeyValue;
+import org.apache.hadoop.hbase.TableName;
+import org.apache.hadoop.hbase.backup.impl.BackupAdminImpl;
+import org.apache.hadoop.hbase.backup.impl.BackupManager;
+import org.apache.hadoop.hbase.client.Admin;
+import org.apache.hadoop.hbase.client.ColumnFamilyDescriptorBuilder;
+import org.apache.hadoop.hbase.client.Connection;
+import org.apache.hadoop.hbase.client.ConnectionFactory;
+import org.apache.hadoop.hbase.client.Put;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.hadoop.hbase.client.Table;
+import org.apache.hadoop.hbase.client.TableDescriptorBuilder;
+import org.apache.hadoop.hbase.io.hfile.HFile;
+import org.apache.hadoop.hbase.io.hfile.HFileContextBuilder;
+import org.apache.hadoop.hbase.testclassification.MediumTests;
+import org.apache.hadoop.hbase.testing.TestingHBaseCluster;
+import org.apache.hadoop.hbase.testing.TestingHBaseClusterOption;
+import org.apache.hadoop.hbase.tool.BulkLoadHFiles;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Category(MediumTests.class)
+@RunWith(Parameterized.class)
+public class TestBackupRestoreWithModifications {
```

Comment:
@rubenvw-ngdata Rather than include this test class (which I will commit with my other PR), could you take a look at TestIncrementalBackup? I see there is a `TestIncBackupRestore` method in there, which appears to cover restoring incremental to a new table. I wonder why that test succeeds, and whether it can easily be fixed to cover this case as part of your PR here

Reply:
Actually I think the reason is because in the tableName passed in is only used for setting up the HFileInputFormat2 directories/splits. In the TestIncrementalBackup class, the original table exists when the incremental restore happens. So the test succeeds.

Reply:
@bbeaudreault I have difficulties running the TestIncrementalBackup locally. It must be something timing related. A colleague of mine only gets through it when adding a bunch of breakpoints and debugging.

On the other hand, I think that it is a valid test that is added here with the removeTablesBeforeRestore parameter (willing to rebase it onto your changes once merged).

If you want me to dig deeper into it, A README description of what is required to run everything would really help me.
Without that, it's easier for me to add a standalone test of which I am certain it runs.

Reply:
I just merged HBASE-28456 so this PR now has conflicts. I'd be happy to get this into the 2.6.0 release if we can resolve it shortly. I don't have a strong opinion on where the test lives, though in the context of this jira/PR I'd prefer to make it clear that we're testing "select correct table for incremental backup" rather than "modifications on table for incremental backup" (the previous jira).

Would you mind clearing that up? You also need to run `mvn spotless:apply -pl hbase-backup`. We should be able to merge quickly once these are resolved.

Reply:
@rubenvw-ngdata sorry I did not see your response before I posted my response above. Can you create a new test class, like TestIncrementalRestoreNewTable or something with just this new test for this particular case? It's a little duplicative but is probably the clearest thing without diving into the much larger TestIncrementalBackup.

Does that sound reasonable?

Reply:
Just confirming that the test has timing related issues at my end to. We're trying to find out where the timing issue is.

What I'll do for this ticket, I'll create a separate test for restoring a backup (full or incremental) on an empty and non empty environment

Reply:
Sounds great, thanks! I'll keep an eye out.

Reply:
@rubenvw-ngdata one last thing -- i know it's getting late over there in Ghent (based on your profile), but I'd like to start the 2.6.0 release tomorrow morning my time (Eastern Time). Happy to include this if we can get the test fixed up before then. No pressure, just letting you know of my timeline and acknowledging the timezone differences. Thanks again.

Reply:
@bbeaudreault should be good now. I'm still here for 1 hour if there is anything left to change. Tomorrow, I'm not working, so hope it is ok :)